### PR TITLE
Improve formatting tags

### PR DESCRIPTION
**Pull request by Jo-Be-Co**
* Consistent formatting options for all text elements – including placeholders such as `{T}`
* Text can be displayed in uppercase, lowercase or title case. It is also possible to truncate text to a maximum length.
* Multi-level sorting of lists
* Reverse sorting
* Regular expressions with named capture groups and comments
* Escape option in formats
* Content validation at the `<has->` tag – though within a new `<is->` tag
* Text checks and numerical checks also on individual elements of a list
* New `<minutes>` tag
* New list tag `<tag>`
* New conditional tag `<if abridged->`

### CODE REVIEW

**Comment by Jo-Be-Co** [comment]
Let me know if you would like to import only some parts of this or if I could improve some lines.

I split my work into smaller commits so it’s easier to review and understand.

**Comment by rmcrackan** [comment]
First thing's first -- this is incredible work! Thank you so much. Some of this is simple evolution of what's there and other stuff is all new territory. Specifically regex support. Love the comments/doc.s also. Did you write this by hand or is this AI? Either way it's quite good. My items below are mostly minor polish, safety hardening, and test gaps.

Full disclosure: I used AI to help write this review. I know it's lengthy but **I read, edited, and stand by every word**. Honestly I probably read/edited it a dozen times to make sure it's not just generated slop.

---

### What went well

* `TemplatesTests` and `FileNamingTemplateTests` were expanded meaningfully (minutes, slice/sort, many `<is>` cases, abridged, tags, culture).
* `docs/features/naming-templates.md` updates. No one likes keeping the docs updated but this is the only way users can know, and I'm really thankful.
* Centralizing formatters and threading `CultureInfo` through the expression tree is an improvement over ad hoc string replacement for contributors and similar cases.

---

### Issues to address

* **Thread safety: `CultureInfo.CurrentCulture` in regex checks**  
In `ConditionalTagCollection`, the `~` (regex) path temporarily assigns `CultureInfo.CurrentCulture` while running `Regex.IsMatch`. That is **process-global mutable state** and is risky if template evaluation ever runs concurrently (or overlaps with other code that assumes a stable current culture). Prefer passing culture into `Regex` via a pattern that does not depend on current culture, or use overloads / `RegexOptions.CultureInvariant` where appropriate, without mutating `CurrentCulture`. I can't think of a use case where this could happen in practice and it's not a hill I'm going to die on. It'd be nice to fix it now though.

* **Malicious regex vulnerability**  
Hopefully I'm not borrowing trouble here, but I could imagine someone asking for help, esp on reddit, and a malicious user giving them poisoned regex. Or of course, it could even be a simple fuck up like `(a+)+` or `(a|aa)+`. Anyway, the point is: a catastrophic backtracking pattern in `<is ... [~...]>` could **peg CPU** (ReDoS-style). Please consider imposing a timeout and a clear error message.

* **Build / analyzer noise**  
The PR introduces several nullable reference warnings in `ConditionalTagCollection` and delegate nullability mismatches in tests. Cleaning these up improves maintainability.

* **Documentation nits**  
    * Minutes section links to `#number-formatter` but the heading is `#number-formatters` (anchor may be broken).  
    * A few typos remain ("Speficy", "Form more", "correspoing").  
    * One text-list `sort` example result looks suspicious (duplicate token in the displayed result—worth verifying against the test data).

---

### Test requests

* **`<!is ...>`** is documented in `naming-templates.md` but I did not find matching tests.
* **Escaped strings** in `<is ...>` checks and in `[...]` format segments (per doc: backslash escaping, including `]`).
* **`NamingTemplate.Evaluate` argument binding** when **multiple arguments** could satisfy the same parameter type (guards the new `FirstOrDefault(IsInstanceOfType)` behavior). This one is odd. I'm actually going to address it specifically in my next comment since this review has already gotten unreasonably long.
* **Focused `CommonFormatters` tests** for edge cases (unknown `{tag}` in contributor-style templates, minutes boundaries, invalid combined format strings).

**Comment by rmcrackan** [comment]
Explanation of the above 'arg binding when mult args':

We want to encode "when I pass these 2 objects and the delegate expects these 2 types, I get *this* mapping." So that if someone later changes the algo (positional join, consume-used-args, throw on ambiguity), then I'll see it when regression tests fail. 

Specific call-outs:

* **2 parameters with the same type**  
If the compiled delegate had 2 parameters both typed as e.g. `LibraryBookDto`, the loop would run twice and each time `FirstOrDefault` would return the same 1st matching object. You never bind the 2nd DTO to the 2nd parameter. So you can't support "2 books" by passing 2 instances -- unless the delegate's parameter types differ.

* **1 runtime object matches more than 1 parameter type**  
Example: parameters `(object a, SpecificDto b)` and you pass `[specificDto]`. The 1st parameter might bind to that instance (as `object`), and depending on order of parameters, the 2nd might also see it as assignable -- or you might get wrong pairing if you had a weirder type layout. The subtle case is really **duplicate parameter types** or **everything passed as a common base/interface** so "1st match" isn't unique.

* **order vs 1st fit**  
If you pass `[dtoB, dtoA]` but the delegate expects `(DtoA, DtoB)`, you might still get the right binding **only if** each type appears once and the 1st occurrence in your array is the one you want. If both are compatible with the 1st parameter's type, ordering surprises appear.

**Comment by Jo-Be-Co** [comment]
Thanks for the initial, yet already quite detailed, review. I’m sure I can sort most of this out fairly quickly. As for the other points, I’ll have to see if I can manage them.

Actually, this is the first C# project I’ve worked on. And I’m constantly surprised by the shortcuts the language allows in certain places.
As I usually use a JetBrains IDE, I installed RIDER here and got to work on the code. Most of it was indeed written by hand – hence the odd typo or two.
As I wasn’t entirely familiar with expressions (at the start), I actually had the AI implement the CultureInfo parameter in a first draft. In doing so, the argument handling was also modified to incorporate this parameter.
The already quite comprehensive test suite did a great job here, in my view. But the corner cases you listed haven’t (yet) been spotted, of course.

PS: I created all the regular expressions by hand. At this stage, I haven’t yet figured out how to work with timeouts here. However, I don’t see the injection risk as particularly significant, as the user creates the formats for execution on their own system.

**Comment by Jo-Be-Co** [comment]
BTW: many of the refactorings where just suggestions by RIDER. And I just wanted to get the warning level low on the files I have been working on.

**Comment by rmcrackan** [comment]
> this is the first C# project I’ve worked on

Woah, seriously? You must have a lot of experience in other languages. This is definitely not novice work :)

> I’m constantly surprised by the shortcuts the language allows in certain places

C# is hands-down my favorite language. It's not flawless, but it does so so **so** many things right.

> I don’t see the injection risk as particularly significant, as the user creates the formats for execution on their own system.

I agree it's their own fault and their own machines, but if the user screws the pooch hard bad enough -- I'm the one who has to address their hung system in a panicked trouble ticket.

IsMatch has an overload with timeout, so something like this should be possible (this is untested):

    private const int REGEX_TIMEOUT_MS = 250;
    private static TimeSpan TemplateRegexTimeout => TimeSpan.FromMilliseconds(REGEX_TIMEOUT_MS);
    
    try
    {
        return Regex.IsMatch(
            v?.ToString().Trim() ?? "",
            valStr,
            RegexOptions.IgnoreCase | RegexOptions.CultureInvariant,
            TemplateRegexTimeout);
    }
    catch (RegexMatchTimeoutException ex)
    {
        throw new InvalidOperationException(
            "Naming template: a regex in an <is ... [~...]> check took too long and was stopped. " +
            "Use a simpler pattern or remove that condition. " +
            $"Pattern starts with: {Truncate(valStr, 100)}",
            ex);
    }

Small but important tweak: I dropped the CultureInfo.CurrentCulture assignment around IsMatch and I instead used RegexOptions.CultureInvariant in my example. That removes the thread-safety footgun the review called out and is still consistent with "case-insensitive match on user text" for most title/author checks.

**Comment by Jo-Be-Co** [comment]
> > this is the first C# project I’ve worked on
> 
> Woah, seriously? You must have a lot of experience in other languages. This is definitely not novice work :)

Yes. Have only been used to Java, Perl, Shell and some Python ...

> > I don’t see the injection risk as particularly significant, as the user creates the formats for execution on their own system.
> 
> I agree it's their own fault and their own machines, but if the user screws the pooch hard bad enough -- I'm the one who has to address their hung system in a panicked trouble ticket.

Understandable. The faults which could be done by users will be done. 😉
 
> IsMatch has an overload with timeout, so something like this should be possible (this is untested):

Didn't spot your reply until now. So I came to a likewise solution. I might introduce a constant. And I’ve gone and removed all that fiddling around with CurrentCulture as well. Regex simply pulls it straight from the environment, but the supposed benefit is minimal. I’d gone through all the text methods almost fanatically, checking for CultureInfo, but with Regex, the only thing that worked was the big hammer...

> ### Test requests
> * **`NamingTemplate.Evaluate` argument binding** when **multiple arguments** could satisfy the same parameter type (guards the new `FirstOrDefault(IsInstanceOfType)` behavior). This one is odd. I'm actually going to address it specifically in my next comment since this review has already gotten unreasonably long.
> * **Focused `CommonFormatters` tests** for edge cases (unknown `{tag}` in contributor-style templates, minutes boundaries, invalid combined format strings).

These are still missing though.

**Comment by rmcrackan** [comment]
This is lovely work.

I'm on the fence about swallowing regex errors and returning false. But instead of holding up your work, I'll accept the PR now; those edge cases can be addressed later if it's important.

Thank you again!

**Comment by rmcrackan** [comment]
Spoke too soon. Looks like some failing unit tests are blocking the merge. Let me know when those are cleaned up.

**Comment by Jo-Be-Co** [comment]
I've also changed the regexp handling to throw exceptions. As well as some test to trigger this.
So this is waiting in my IDE to become a commit 😉

**Comment by Jo-Be-Co** [comment]
Ok - just spotted the failed tests. This is a speciality with different OSes.
On my windows environment they passed ... will fix this!

**Comment by Jo-Be-Co** [comment]
@rmcrackan
now you might trigger another build. On my system the unit tests don't show errors.

**Comment by rmcrackan** [comment]
Much thanks for the quick turn around. I'll take another look very soon.

**Comment by rmcrackan** [comment]
I added 2 minor notes

File: Source/FileManager/NamingTemplate/ConditionalTagCollection[TClass].cs [hunk 1]
```diff
@@ -202,55 +204,78 @@ private static ConditionEvaluator GetPredicate(string? checkString)
 		private static int VComparedToStr(object? v, CultureInfo? culture, string valStr)
 		{
 			culture ??= CultureInfo.CurrentCulture;
-			return culture.CompareInfo.Compare(v?.ToString()?.Trim(), valStr, CompareOptions.IgnoreCase);
```
[the remark this answers is not in the document]
Comment (rmcrackan) [reply Source/FileManager/NamingTemplate/ConditionalTagCollection[TClass].cs]: Sounds good to me

**Comment by rmcrackan** [comment]
Ready when you are -- I'm very happy with how things are. Do you have some more tweaking to do? Or should I merge?

**Comment by Jo-Be-Co** [comment]
I just thought somewhat more about the Trim() at the comparisons ...
If I want to compare with `[ = test ]` this should be completely shrunken to just `test`.
Having a value from audible which contains space at either end or if I cut it right after a space this might be hard to compare.

So `<is title[4][=the ]->` might be unmatchable.

Maybe escaping the space after "the" might be an alternative to Trim(). But anyway, there seems to be a test missing.

**Comment by Jo-Be-Co** [comment]
One question from my side: are you fine with all the syntax and naming for the patterns I came up with?

- slice()
- numerical checks with `#=` and more
- `<is foobar[format][check]->`
- title case

in my first attempt i just extended the `has` with checks. This is quite identical to the other check but having only a simple null check.

**Comment by rmcrackan** [comment]
> <is title[4][=the ]->

Hmm, I see what you mean. Maybe Trim() won't work. I guess we should just treat that as literal. You included a space? You get a space!

> are you fine with all the syntax and naming for the patterns I came up with?

These all look good to me.

Thoughts?

**Comment by Jo-Be-Co** [comment]
> > <is title[4][=the ]->
> 
> Hmm, I see what you mean. Maybe Trim() won't work. I guess we should just treat that as literal. You included a space? You get a space!

I think the Trim() might have made it work as a "The " would have been trimmed to "The" and the comparison from "[=the ]" which would have matched "=the" as check-code. This goes down to a case insensitive `"The" == "the"`.

I definitely would add some tests.

But which of these checks shall match? Given a title of `"The ABC murders"`:
* `<is title[4][=the]->`
* `<is title[4][=the ]->`
* `<is title[4][=the\ ]->` ✅
* `<is title[4][= the]->`
* `<is title[4][= the ]->`
* `<is title[4][= the\ ]->` ✅
* `<is title[4][ =the]->`
* `<is title[4][ =the ]->`
* `<is title[4][ =the\ ]->` ✅

As you stated "You included a space? You get a space!". This would somehow be opposite to the handling of the other parsing. `<first author[  {F} {L]  ]>` won't get all of the specified spaces.

Please do let me know what you think, and I’ll incorporate your feedback accordingly. Otherwise, I would initially implement the following version:
* No `Trim()` on `title[4]`
* Spaces after the `[` and before the `]` are not included.
* Spaces after the operator are not included.
* Spaces can be included by using an escape character.

> > are you fine with all the syntax and naming for the patterns I came up with?
> 
> These all look good to me.
> 
> Thoughts?

At what are you aiming? If my proposal is acceptable, it would be fine.

**Comment by Jo-Be-Co** [comment]
Ok, let's get this deployed 🚀

**Comment by rmcrackan** [comment]
Yeah, I see your point.

Is our only hang-up about whitespace within `<is `? If yes, what about handling trim differently for `is` (and `!is`) vs everything else?

`is` is a boolean return (I hated having to type 'is is' lol), which is fundamentally different than a formatter. ie: one verifies, the other performs an action. Or put another way, the formatter actually produces data, whereas `is` returns metadata. It seems reasonable that a verifier and a mutator would handle the same input differently.

Also, by just splitting the paths for `is` vs the rest, I think we can avoid all of the nuanced specific cases you listed. RE whitespace around brackets, operators, etc.

Thoughts?

**Comment by rmcrackan** [comment]
Also: I cannot thank you enough for commenting that monster regex. Regex is so easy to write and so hard to read. That would have been impossible to grasp without your comments.

**Comment by rmcrackan** [comment]
Yet another addendum: thanks for working the problem with me. This is way more fun than working alone :)

**Comment by Jo-Be-Co** [comment]
I have to check at which locations to have these "ignore whitespace" implementations.

There have been some. Mostly it has been about RE, taking the match and trimming it.

Where trimming was used I tuned the REs so that
- the whole match could be taken unchanged
- a match would only occur with usable content
- thus matches and/or groups could be queried with `.Success` independent from contained characters

There were also unit tests explicitly playing around with additional whitespace. So for me this has been the rule to follow ...

I came up with the use of escapings just because I like to use regular expressions (who would have thought) and I saw here the problem of wanting to enter square brackets within square brackets.

I could create an overview tonight (Europe) of where spaces are now handled.

**Comment by Jo-Be-Co** [comment]
> Yet another addendum: thanks for working the problem with me. This is way more fun than working alone :) 

For me, this is definitely the most fun way to program. Unfortunately, pair programming takes place far too rarely, but in connection with code quality, you can't get around pull requests anymore. And then that's exactly the time when you should still include improvements. Everything else ends only in future tasks and woulda, coulda, shoulda.

**Comment by Jo-Be-Co** [comment]
> Also: I cannot thank you enough for commenting that monster regex. Regex is so easy to write and so hard to read. That would have been impossible to grasp without your comments.

One great thing about Regex is its compactness. But you find yourself counting brackets when you need to identify capture groups. So I often use non-capturing groups, which adds more characters and also names the capturings, which makes it even longer.

If things get too weird, you simply need a way to comment on these monsters and break up these wow beasts into understandable chunks.

If they don't work and you have to recapture yourself of what you did and why you did it... in that situation, you also learn that commenting is not only for the rest of the world 🤓

**Comment by rmcrackan** [comment]
[Since I assume this will be resolved](https://github.com/rmcrackan/Libation/commit/0960edeafb42d82e01634f615575f5d1566d01f1) :)

**Comment by Jo-Be-Co** [comment]
So here is another one. I just rechecked all places where whitespace from input might be skipped or not. And I also wondered where else escaping might be needed. The `separator()` and `format()` commands caught my eye in this regard. But I decided to hold off on escaping for the time being. The `<format>` tag at least allows for this.
If I discard fewer spaces now, I’ll end up with “You included a space? You get a space!”. This just changes some test outcomes on my freshly added tests for this question.
When it came to number formats, I also wondered whether escaping might be necessary here. This then raised the question of which characters are actually allowed in a number format, and whether they might conflict with the respective terminating character? `<bitrate[format]>` or `<series[format{#:format}]>`.
At least, based on the documentation, one might get the impression that [custom numeric format strings](https://learn.microsoft.com/en-us/dotnet/standard/base-types/custom-numeric-format-strings) are fully supported.
It then occurred to me that strings enclosed in `'...'` or `"..."` are not taken into account. At least, they do not work if the respective closing bracket is included.
**Pimp my regular expression:** So I corrected that!

**Comment by Jo-Be-Co** [comment]
Since numbers and date values already make use of .NET’s formatting options, I’m wondering whether minutes should also be handled as a timespan [in this way](https://learn.microsoft.com/en-us/dotnet/standard/base-types/custom-timespan-format-strings).
The tag is still new...

**Comment by Jo-Be-Co** [comment]
I guess I’ll have to look at another mistake.

In the tests I had closed several `is` tags with `has`, by Copy & Paste. There was no difference to correctly closed tags ...

**Comment by rmcrackan** [comment]
Sorry I was absent yesterday. Work was superlatively nuts and when I finally got free time in the evening ngl I was burnt out as hell from the day. I'm back for real now.

I thought of another annoying exception to 'You included a space? You get a space!' -- spaces around blank entries:  
`{F} {M} {L}` obviously doesn't want 2 spaces when the middle name is absent. I feel like that was addressed previously but you're the one in the guts lately so frankly you'd know better than me right now. It's almost like it needs multiple passes:

---

`{F} {M} {L} ({S})` : Bruce [no middle name] Wayne [no suffix]
null/empty/whitespace template entries actually apply a null char (\0):
=> `Bruce \0 Wayne (\0)`

then, pattern:

      \s\0\s    # single space NUL single space
    | ^[\0\s]+  # starts with NULs and spaces
    | [\0\s]+$  # ends with NULs and spaces

replacement: " "
=> `Bruce Wayne (\0)`

then get rid of the remaining nulls
replace `\0` => ""
=> `Bruce Wayne ()`

---

Again, this might already be addressed. If so, I'm sure it's in the tests.

**Comment by rmcrackan** [comment]
> No made up format for minutes as .NET already knows how to format them.

Good call

**Comment by Jo-Be-Co** [comment]
> Sorry I was absent yesterday. Work was superlatively nuts and when I finally got free time in the evening ngl I was burnt out as hell from the day. I'm back for real now.
> 
> I thought of another annoying exception to 'You included a space? You get a space!' -- spaces around blank entries: `{F} {M} {L}` obviously doesn't want 2 spaces when the middle name is absent. [...]
> `{F} {M} {L} ({S})` : Bruce [no middle name] Wayne [no suffix] null/empty/whitespace template entries actually apply a null char (\0): => `Bruce \0 Wayne (\0)`

The [TemplateStringFormatter](https://github.com/rmcrackan/Libation/pull/1690/changes#diff-2d5da1b450ab4ea1076491fb9d7167fb521ee777591c797bd1aef5adbe49aeb4R59) will replace every single placeholder by it's value applying an optional format.

The used [_StringFormatter](https://github.com/rmcrackan/Libation/pull/1690/changes#diff-2d5da1b450ab4ea1076491fb9d7167fb521ee777591c797bd1aef5adbe49aeb4R37) explicitly sets the value to `string.Empty`. So no `\0` will be inserted.

At last multiple spaces will be compacted and the result will be trimmed by [CollapseSpacesAndTrimRegex](https://github.com/rmcrackan/Libation/pull/1690/changes#diff-2d5da1b450ab4ea1076491fb9d7167fb521ee777591c797bd1aef5adbe49aeb4R65)

> Again, this might already be addressed. If so, I'm sure it's in the tests.

Hopefully addressed in a proper way. The [tests using the author tag](https://github.com/rmcrackan/Libation/pull/1690/changes#diff-f1ff47c314ea6d9e010f1767c3c91d57aaf30198060681e74efd96d80d2d39bcR383) where no format is specified will use [the default format](https://github.com/rmcrackan/Libation/pull/1690/changes#diff-97ec8fbc03457046e7651bb0c5a9c4370ed99b10225dbb3c17865677b6d6cdfcR26). All these tests expect spaces to be reduced correctly.

**Comment by Jo-Be-Co** [comment]
> Sorry I was absent yesterday. Work was superlatively nuts and when I finally got free time in the evening ngl I was burnt out as hell from the day. I'm back for real now.

Happy to have you back here. Sometimes you just have days like that…

**Comment by Jo-Be-Co** [comment]
> I guess I’ll have to look at another mistake.
> 
> In the tests I had closed several `is` tags with `has`, by Copy & Paste. There was no difference to correctly closed tags ...

This is already addressed by [some tests](https://github.com/rmcrackan/Libation/pull/1690/changes#diff-9d113341ff290a984ee503cb0ed5d0b106839b86124a90fd37f86aad0f569097R229)

**Comment by Jo-Be-Co** [comment]
The new TimeSpan tag is implemented and documented end‑to‑end. Its design keeps backward compatibility in mind, so future tweaks shouldn’t break existing formats. I do have a few follow‑up ideas, but I’m happy to call this PR complete for now — open to feedback or requested changes. 😄

**Comment by rmcrackan** [comment]
> I’m happy to call this PR complete for now

Sounds good to me. Was there ever a solution to the `=the ` issue? Or is that just going to be an outstanding edgecase for now?

**Comment by rmcrackan** [comment]
I really do want to merge this. I swear I'm not trying to find reasons not to. But something about it is killing the validation step : https://github.com/rmcrackan/Libation/actions/runs/23624513072

Fortunately, it looks like just 1 unit test.

**Comment by Jo-Be-Co** [comment]
> > I’m happy to call this PR complete for now
> 
> 
> 
> Sounds good to me. Was there ever a solution to the `=the ` issue? Or is that just going to be an outstanding edgecase for now?

`=the ` issue seized to be. This has been addressed with the decission to keep spaces as their are.
* `<is title[4][=the]->`
* `<is title[4][=the ]->` ✅
* `<is title[4][=the\ ]->` ✅
* `<is title[4][= the]->`
* `<is title[4][= the ]->`
* `<is title[4][= the\ ]->`
* `<is title[4][ =the]->`
* `<is title[4][ =the ]->` ✅
* `<is title[4][ =the\ ]->` ✅

[tested with =An ](https://github.com/rmcrackan/Libation/blob/dd2bc7c461773743a740f6717ce4d09b8ea00143/Source/_Tests/LibationFileManager.Tests/TemplatesTests.cs#L566)

**Comment by Jo-Be-Co** [comment]
> I really do want to merge this. I swear I'm not trying to find reasons not to. But something about it is killing the validation step : https://github.com/rmcrackan/Libation/actions/runs/23624513072
> 
> 
> 
> Fortunately, it looks like just 1 unit test.

In a way it is my fault ...

For this case the test depends on a specific internationalization.

**Comment by rmcrackan** [comment]
Hallelujah! Release coming soon. Thanks for sticking with this.
